File: src/de/esymetric/jerusalem/utils/RandomAccessFileCache.java
package de.esymetric.jerusalem.utils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

public class RandomAccessFileCache {

	final static int MAX_SIZE = 200; // was 256

	Queue<String> queue = new LinkedList<String>();
	Map<String, RandomAccessFile> cache = new HashMap<String, RandomAccessFile>();

	int numberOfAccesses;
	int numberOfFileChanges;

	public RandomAccessFile getRandomAccessFile(String filePath,
			boolean readOnly) {
		numberOfAccesses++;

		// get it from cache?

		if (cache.containsKey(filePath)) {
			queue.remove(filePath);
			queue.add(filePath); // insert at end of queue
			return cache.get(filePath);
		}

		// not in queue

		if (readOnly && !new File(filePath).exists())
			return null;

		// remove first element?

		if (queue.size() >= MAX_SIZE) {
			String fp = queue.poll();
			RandomAccessFile raf = cache.remove(fp);
			try {
				raf.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		RandomAccessFile raf = null;
		try {
			raf = new RandomAccessFile(filePath, readOnly ? "r" : "rw");
			numberOfFileChanges++;
		} catch (FileNotFoundException e) {
			e.printStackTrace();
			System.out
					.println("FAILURE to open RandomAccessFile, file cache size: "
							+ cache.size());
		}

		if (raf == null) {
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
			}

			new File(filePath).delete();

			try {
				raf = new RandomAccessFile(filePath, readOnly ? "r" : "rw");
				numberOfFileChanges++;
			} catch (FileNotFoundException e) {
				e.printStackTrace();
				System.out
						.println("FAILURE(2) to open RandomAccessFile, file cache size: "
								+ cache.size());

				return null;
			}
		}

		cache.put(filePath, raf);
		queue.add(filePath);

		return raf;
	}

	public int getAndClearNumberOfFileChanges() {
		int n = numberOfFileChanges;
		numberOfFileChanges = 0;
		return n;
	}

	public void close() {
		for (RandomAccessFile raf : cache.values())
			try {
				raf.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		queue.clear();
		cache.clear();
	}
}

File: src/main/java/de/esymetric/jerusalem/utils/RandomAccessFileCache.kt
package de.esymetric.jerusalem.utils

import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.RandomAccessFile
import java.util.*

class RandomAccessFileCache {
    var queue: Queue<String> = LinkedList()
    var cache: MutableMap<String, RandomAccessFile> = HashMap()
    var numberOfAccesses = 0
    var numberOfFileChanges = 0
    fun getRandomAccessFile(
        filePath: String,
        readOnly: Boolean
    ): RandomAccessFile? {
        numberOfAccesses++

        // get it from cache?
        if (cache.containsKey(filePath)) {
            queue.remove(filePath)
            queue.add(filePath) // insert at end of queue
            return cache[filePath]
        }

        // not in queue
        if (readOnly && !File(filePath).exists()) return null

        // remove first element?
        if (queue.size >= MAX_SIZE) {
            val fp = queue.poll()
            val raf = cache.remove(fp)
            try {
                raf!!.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        var raf: RandomAccessFile? = null
        try {
            raf = RandomAccessFile(filePath, if (readOnly) "r" else "rw")
            numberOfFileChanges++
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
            println(
                "FAILURE to open RandomAccessFile, file cache size: "
                        + cache.size
            )
        }
        if (raf == null) {
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
            }
            File(filePath).delete()
            try {
                raf = RandomAccessFile(filePath, if (readOnly) "r" else "rw")
                numberOfFileChanges++
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                println(
                    "FAILURE(2) to open RandomAccessFile, file cache size: "
                            + cache.size
                )
                return null
            }
        }
        cache[filePath] = raf
        queue.add(filePath)
        return raf
    }

    val andClearNumberOfFileChanges: Int
        get() {
            val n = numberOfFileChanges
            numberOfFileChanges = 0
            return n
        }

    fun close() {
        for (raf in cache.values) try {
            raf.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        queue.clear()
        cache.clear()
    }

    companion object {
        const val MAX_SIZE = 200 // was 256
    }
}